examples/interactive_asr/vad.py

```diff
@@ -0,0 +1,278 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+#
+# This source code is licensed under the license found in the LICENSE file in
+# the root directory of this source tree. An additional grant of patent rights
+# can be found in the PATENTS file in the same directory.
+"""
+Following `a simple but efficient real-time voice activity detection
+algorithm <https://www.eurasip.org/Proceedings/Eusipco/Eusipco2009/contents/papers/1569192958.pdf>`__.
+"""
+
+from collections import deque
+
+import librosa
+import numpy as np
+import torch
+
+import pyaudio
+import torchaudio
+from six.moves import queue
+
+######################################################################
+# There are three criteria to decide if a frame contains speech: energy,
+# most dominant frequency, and spectral flatness. If any two of those are
+# higher than a minimum plus a threshold, then the frame contains speech.
+# In the offline case, the list of frames is postprocessed to remove too
+# short silence and speech sequences. In the online case here, inertia is
+# added before switching from speech to silence or vice versa.
+#
+
+
+def compute_spectral_flatness(frame):
+    EPSILON = 0.01
+    n = frame.nonzero().size(0)
+    geometric_mean = torch.exp((EPSILON + frame).log().mean(-1)) - EPSILON
+    arithmetic_mean = frame.mean(-1)
+    return -10 * torch.log10(EPSILON + geometric_mean / arithmetic_mean)
+
+
+def compute_energy(frame):
+    return frame.pow(2).sum(-1)
+
+
+class VoiceActivityDetection(object):
+    def __init__(
+        self,
+        num_init_frames=30,
+        ignore_silent_count=4,
+        ignore_speech_count=1,
+        energy_prim_thresh=60,
+        frequency_prim_thresh=10,
+        spectral_flatness_prim_thresh=3,
+        verbose=False,
+    ):
+
+        self.num_init_frames = num_init_frames
+        self.ignore_silent_count = ignore_silent_count
+        self.ignore_speech_count = ignore_speech_count
+
+        self.energy_prim_thresh = energy_prim_thresh
+        self.frequency_prim_thresh = frequency_prim_thresh
+        self.spectral_flatness_prim_thresh = spectral_flatness_prim_thresh
+
+        self.verbose = verbose
+
+        self.speech_mark = True
+        self.silence_mark = False
+
+        self.silent_count = 0
+        self.speech_count = 0
+        self.n = 0
+
+        if self.verbose:
+            self.energy_list = []
+            self.frequency_list = []
+            self.spectral_flatness_list = []
+
+    def iter(self, frame):
+
+        EPSILON = 1.0
```

Comment:
we could use sys.float_info.epsilon for EPSILON here and in compute_spectral_flatness?

Reply:
move that `EPSILON` within the one place where it is used. there's another epsilon, with a different value, that was renamed to avoid confusion. the model is sensitive to the values of both, and were calibrated.